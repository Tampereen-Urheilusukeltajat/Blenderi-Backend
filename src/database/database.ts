import config from '../../knexfile';
import knex from 'knex';

export enum Env {
  development = 'development',
  production = 'production',
}

const ENV = (process.env.ENVIRONMENT as Env) ?? Env.development;

export const knexController = knex(config[ENV]);
