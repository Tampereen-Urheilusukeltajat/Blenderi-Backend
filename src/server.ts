import fastify, { FastifyInstance } from 'fastify';
import { fastifySwagger } from '@fastify/swagger';
import { fastifyHelmet } from '@fastify/helmet';
import fastifyCors from '@fastify/cors';
import { fastifyAutoload } from '@fastify/autoload';
import { TypeBoxTypeProvider } from '@fastify/type-provider-typebox'

import { log } from './lib/log';
import path from 'path';

export const buildServer = async (opts: {
  routePrefix: string
}): Promise<FastifyInstance> => {
  const server = fastify({
    logger: false,
    ignoreTrailingSlash: true,
    ajv: {
      customOptions: {
        removeAdditional: 'all' // Remove additional params from the body etc
      }
    }
  });

  // Register plugins and routes
  server
    .addSchema({
      $id: 'error',
      type: 'object',
      properties: {
        statusCode: { type: 'number' },
        error: { type: 'string' },
        message: { type: 'string' }
      }
    })
    .register(fastifySwagger, {
      routePrefix: `${opts.routePrefix}/documentation`,
      swagger: {
        info: {
          title: 'Blenderi REST API',
          description: 'Documentation for Blenderi REST API',
          version: '0.0.1'
        },
        securityDefinitions: {
          bearerAuth: {
            type: 'apiKey',
            name: 'Authorization',
            in: 'header'
          }
        },
        security: [{
          bearerAuth: []
        }],
        consumes: ['serverlication/json'],
        produces: ['serverlication/json'],
        tags: [{
          name: 'Utility',
          description: 'Utility endpoints'
        }
        ]
      },
      exposeRoute: true
    })
    .register(fastifyHelmet, {
      contentSecurityPolicy: {
        directives: {
          defaultSrc: ['\'self\''],
          styleSrc: ['\'self\'', '\'unsafe-inline\''],
          imgSrc: ['\'self\'', 'data:', 'validator.swagger.io'],
          scriptSrc: ['\'self\'', 'https: \'unsafe-inline\'']
        }
      }
    })
    .register(fastifyCors, {
      origin: (origin, cb) => {
        // TODO: Configure CORS
        cb(null, true);
      }
    })
    .register(fastifyAutoload, {
      dir: path.join(__dirname, 'routes'),
      dirNameRoutePrefix: (_folderParent, folderName) => `${opts.routePrefix}/${folderName}`
    })
    .setErrorHandler(async (error, request, reply) => {
      log.error({
        error: error.name,
        message: error.message,
        url: request.url,
        method: request.method,
        body: request.body,
        stack: error.stack
      });

      await reply.status(500).send({
        statusCode: 500,
        error: 'Internal Server Error',
        message: 'Internal Server Error'
      });
    })
    .withTypeProvider<TypeBoxTypeProvider>();

  return server;
};
