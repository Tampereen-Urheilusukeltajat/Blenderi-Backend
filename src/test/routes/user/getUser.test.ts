/* eslint-disable @typescript-eslint/no-unsafe-argument */
import {
  describe,
  test,
  expect,
  beforeAll,
  afterAll,
  beforeEach,
} from '@jest/globals';
import { type FastifyInstance } from 'fastify';
import { knexController } from '../../../database/database';
import { buildServer } from '../../../server';
import {
  createTestDatabase,
  dropTestDatabase,
  startRedisConnection,
  stopRedisConnection,
} from '../../../lib/utils/testUtils';

describe('Get user', () => {
  const getTestInstance = async (): Promise<FastifyInstance> =>
    buildServer({
      routePrefix: 'api',
    });

  beforeAll(async () => {
    await createTestDatabase('get_user');
    await startRedisConnection();
  });

  afterAll(async () => {
    await dropTestDatabase();
    await knexController.destroy();
    await stopRedisConnection();
  });

  let server;
  let headers: object;
  beforeEach(async () => {
    server = await getTestInstance();
    const res = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'test@email.fi',
        password: 'password',
      },
    });
    const tokens = JSON.parse(res.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };
  });
  describe('User is found', () => {
    test('it returns 200 and has right properties', async () => {
      const res = await server.inject({
        url: '/api/user/1be5abcd-53d4-11ed-9342-0242ac120002/',
        method: 'GET',
        headers,
      });
      const resBody = JSON.parse(res.body);
      expect(res.statusCode).toEqual(200);
      expect(resBody).toMatchSnapshot();
    });
  });
  describe('User cant be returned', () => {
    test('it returns 404 if user is archived', async () => {
      const res = await server.inject({
        url: '/api/user/fbdfc65b-52ce-11ed-85ed-0242ac120002/',
        method: 'GET',
        headers,
      });
      expect(res.statusCode).toEqual(404);
      const resBody = JSON.parse(res.body);
      expect(resBody).toHaveProperty('error');
      expect(resBody).toHaveProperty('message');
    });

    test('it returns 404 if user is deleted', async () => {
      const res = await server.inject({
        url: '/api/user/db2e9bfa-53db-11ed-9342-0242ac120002',
        method: 'GET',
        headers,
      });
      expect(res.statusCode).toEqual(404);
      const resBody = JSON.parse(res.body);
      expect(resBody).toHaveProperty('error');
      expect(resBody).toHaveProperty('message');
    });
  });
});
