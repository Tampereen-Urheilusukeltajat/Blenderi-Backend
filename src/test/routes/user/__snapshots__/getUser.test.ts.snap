// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get user User is found it returns 200 and has right properties 1`] = `
{
  "email": "test@email.fi",
  "forename": "Tester",
  "id": "1be5abcd-53d4-11ed-9342-0242ac120002",
  "isAdmin": true,
  "isAdvancedBlender": false,
  "isBlender": true,
  "isInstructor": false,
  "isMember": true,
  "phoneNumber": "00001",
  "surname": "Blender",
}
`;
